ls1intum/Artemis
# K6: Add ping to keep websocket alive. Update to new WS-message format. Simulate start-now.

### Checklist

### Motivation and Context
In https://github.com/ls1intum/Artemis/pull/1477, the quiz logic got changed in a way that no messages acknowledging a successful submit via WebSockets are sent any more.
The k6 simulation still relied on these confirmation messages, this PR updates the simulation to the new logic.
Also, a bug was fixed where the server closed the WS-connection as the client sent no heartbeat message.

A new flag `-wqs`, defaulting to `false`, was introduced. If enabled, the simulated client will wait until a message triggered by `start-now` is received and only then the submission process will start.

### Description
The Websocket logic got changed so that the simulated client will wait until a message with the result of the quiz is received. The client will terminate once this message is received or a timeout occurred.

The client sends now a message containing `\n` every 20 seconds to the server in order to indicate to the server that the connection is still used/the client is still alive. This is needed as the time between the last submission of the client and the end of the quiz can be longer than the timeout duration defined in the server. The heartbeat-message resets this timeout, the Angular client also uses a similar behaviour.

The newly added flag `-wqs`, short for `waitQuizStart` will indicate if the client should start with submissions ASAP or should wait until a message indicating the quiz start is received. If the `-wqs` flag is enabled, the start of the quiz with waiting clients can be simulated.

### Steps for Testing
1. Verify the Server API test build plans still work after this change.
2. Login to Artemis
3. Navigate to Course Management and create a new quiz exercise
4. Launch the k6 simulation with the `-wqs` flag
5. Check the log of the server: Once every simulated user logged in, the number of open websocket connections/subscriptions should be higher than the number of simulated users (as every client uses at least 1 connection/subscription)
6. Set the quiz visible, nothing should happen yet
7. Start the quiz, now all client should start sending their submissions
8. Once the quiz is over, the results should be sent out to all clients.
9. Verify that all clients submitted their solutions by checking the number of participations in the statistics tab.
